text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,464 @@
+
+This is done in the [`upgradeAssemblyManifest`](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/cx-api/lib/versioning.ts#L60) function.
+
+We stipulate that this is no longer needed. The compatibility checks will make sure we don't add a required property,
+or do any change that enables the CLI to rely on the structure of the new schemas. This will actually also be enforced by
+the compiler itself. For example, if we add a property, it has to be optional (because of the compatibility checks),
+and if its optional, the CLI has to, at compile time, treat the `undefined` scenario, hence, supporting earlier schemas that didn't have this property.
+
+Removing this mechanism will mean that instead of, theoretically, having only place consider older versions of the schema,
+we will now have multiple places in the code that should be aware of it. But thats ok, because its enforced by the compiler,
+and it also provides better context and granularity for doing that so called evolution.
+
+Also, in reality, we don't actually do any evolution. All we do is synthetically upgrade the schema version,
+without adding or modifying any properties.
+
+This brings us to the next item.
+
+##### Compatibility Validation
+
+This is done in the [`verifyManifestVersion`](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/cx-api/lib/versioning.ts#L39) function,
+and contains two validations:
+
+1. **Framework cannot be bigger than CLI.**
+
+   We have already [described](#step-3-validate-cli--framework) how to handle this validation.
+
+2. **CLI cannot be bigger than Framework.**
+
+   This validation is actually misleading, since we do support CLI > Framework. The way its currently implemented is by doing
+   the aforementioned evolution. But, if the evolution is not necessary anymore, than this validation is not necessary anymore.
+   Which makes sense, since we do support this scenario.
+
+### CLI Regression Tests
+
+The second major mechanism we need to implement is CLI regression tests.
+
+As already mentioned, `cx-protocol` API breakage is not the only thing that can go wrong.
+This is to ensure that we don't introduce breaking changes to the CLI itself. The plan here is fairly straight forward:
+
+```console
+git checkout v${LATEST_PUBLISHED_VERSION}
+npm install /path/to/aws-cdk
+npm install /path/to/packages/*
+cd packages/aws-cdk
+yarn integ-cli
+```
+
+> Not to be taken literally, commands just illustrate intent
+
+# Drawbacks
+
+The only drawback I can think of is that we will be forcing users to upgrade their CLI when they upgrade their framework.
+This is an additional action that they didn't have to take until now. However, calling this a *drawback* is debatable,
+since it actually does provide a more consistent behavior than we have now. In any case, we don't consider this a sufficiently serious issue.
```

Comment:
Welll.... this is a drawback in that we already have users complaining that having to touch node stuff (`npm install`, ...) when you're not coding your CDK app in JavaScript/TypeScript feels like an anomaly... now we're proposing to put this hurdle on their way **once a week** (assuming they choose to stay up-to-date)... That's not something we can brush away so easily now!

Reply:
Well, you're right. But is saying that we don't consider it sufficiently serious means we brush it away? I also provide more detailed rationales around this decision below.

Reply:
Our CLI prints an ugly nasty message that encourages people to install the latest version, regardless of the language they use and we have a pretty strong signal that most users will upgrade their CLI *before* they upgrade the framework. CLI upgrades are legitimately perceived as "easier" because they don't interface with user code.

I feel comfortable with this implication given the enormous benefit we get when it comes to compatibility engineering.

Reply:
@RomainMuller 

>  now we're proposing to put this hurdle on their way once a week 

Working on something here. stay tuned.